### ISSUE TITLE
Move Inductor FX passes - pre grad pass to Dynamo

### ISSUE THREAD

**Issue by yanboliang** [pytorch/pytorch#109357]
### 🚀 The feature, motivation and pitch

There are multiple FX passes under ```torch/_inductor/fx_passes``` for various optimization in pre grad pass(torch op level). Though we strongly recommend to put FX passes in post grad, but there are still some for various reasons (E.g patterns can't be recognized in post grad because of autograd/decomp breaks up the patterns). Most of the pre grad passes actually are generic optimization across all backends.

We are thinking about to move them to Dynamo to make all backends benefit from these optimization. We got feature request from different work streams:
* ```torch.export``` should export the optimized graph.
* The [boolean mask rewritten pass](#107844) lives in inductor, but it's basically a requirement for any compiler backend to do something similar.
* Meta internal AOT-Inductor user cases.

This is also relevant to the discussion to move pre grad passes to pre-dispatch passes, maybe we can do the two things together. 

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @jansel @eellison @SherlockNoMad @gmagogsfm

**Comment by jansel** [comment]
I do not like this idea.  I think we should keep graph optimizations out of dynamo.  Dyanmo is a capture mechanism, not an optimizing compiler.  I think we have a good abstraction boundary now, where dynamo just captures the program as written.  Breaking that clean abstraction boundary will be a slippery slope.  There are classes of optimizations that are not always profitable.  There are classes of optimizations that are inductor-specific.  There are classes of optimizations that are bad for some hardware targets, or have complex tradeoffs.

**Comment by gmagogsfm** [comment]
@jansel This idea is largely coming from the existence of pre grad passes on Torch level API. Torch level IR isn't an extension point we want to expose. export() produces ATen-based IR. I think you mentioned that some of the passes can't be moved to so asking users to run FX passes after export() would be hard. 

The other choice would be to not include pre grad passes in OSS export+AoTInductor path

What do you think?

**Comment by jansel** [comment]
> export() produces ATen-based IR

IMO we should set the flag for export to preserve Torch level IR until it get to Inductor.

Lowering to ATen-based IR at export time will subtly break tons of stuff for training, because ATen-IR lowerings depend on grad state.  In other words, make_fx() followed by aot_autograd() is currently broken, and there is not an easy fix.

For export to inference-only edge devices, lowering to ATen-based IR in export makes a lot of sense.  However, if you are eventually just going to just pass it to Inductor it provides zero benefit and just causes problems.

**Comment by gmagogsfm** [comment]
> Lowering to ATen-based IR at export time will subtly break tons of stuff for training, because ATen-IR lowerings depend on grad state. In other words, make_fx() followed by aot_autograd() is currently broken, and there is not an easy fix.

We have plans (after PyTorch Conf) to create a functional pre-dispatch phase IR, which will preserve grad state. It is still ATen-based not Torch API/Module based.

For PyTorch Conf, export + AoTInductor will be inference only.

**Comment by gmagogsfm** [comment]
@jansel I think you mentioned some pre grad passes can't be moved to later phases. I assume the laster phases do not include pre dispatch IR since it doesn't exist yet. Which pass were you referring you? Can they be expressed easily on pre dispatch IR?

**Comment by jansel** [comment]
In theory, all of pre_grad stuff could be moved to some new pre-dispatch IR (which would be different from the existing ATen IR).  In practice, it may take a long time to migrate.

**Comment by gmagogsfm** [comment]
> it may take a long time to migrate.

How come? There doesn't seem to be that many pre_grad passes, seems like 13 of them.

But I do reckon that it won't happen at PyTorch conference, which means we need to make the decision of

1) Whether we include pre_grad passes in the OSS release or not. @yanboliang mentioned there is going to be severe performance penalty for AoTInductor, I didn't get a number though.

2) If we decide to include them, then we would have to create a wrapper-API that guts torch.export to generate Torch level API, run those passes and hide them from users's view. (Until all pre grad passes are moved to pre_dispatch)

**Comment by yanboliang** [comment]
> IMO we should set the flag for export to preserve Torch level IR until it get to Inductor.
> Lowering to ATen-based IR at export time will subtly break tons of stuff for training, because ATen-IR lowerings depend on grad state. In other words, make_fx() followed by aot_autograd() is currently broken, and there is not an easy fix.

This makes sense, I didn't consider too much on training when initializing this idea.

> In theory, all of pre_grad stuff could be moved to some new pre-dispatch IR (which would be different from the existing ATen IR).

I had prototype on this, it works well for my case. But this is a pretty new feature, I can imagine there are more issues if we scale out.

> Whether we include pre_grad passes in the OSS release or not. @yanboliang mentioned there is going to be severe performance penalty for AoTInductor, I didn't get a number though.

I think it should be fine for OSS, but not internal cases. Both AIT and Ads folks did experiments which verified it would have a big perf drop if we remove these pre_grad pass. It makes sense, since most of these passes are inspired from internal cases. So probably we can have the wrapper for internal only now?

**Comment by jansel** [comment]
Do we even need to call export at all?

Can't we just have a `def aot_inductor(fn, args)` that internally calls `export()` with the correct args to keep the Torch IR graph and not call make_fx? This is simpler for the user since there is 1 call rather than 2.

If aot_inductor gets a AtenIR graph, it should just throw an error saying to run `export(aten=False, tracing_mode="none")`, or whatever the flag is to not call `make_fx`.

This seems like a self-inflicted wound.  We are calling `make_fx` right before `aot_autograd` for no reason.  We should just remove the call to `make_fx`.

**Comment by gmagogsfm** [comment]
Having 1 API is the second option I mentioned above, which we can do for
PTC. I do want to ensure the underlying problem, ie. "We should move
pregrad passes to Aten IR" is tracked and addressed some day though.

**Comment by jansel** [comment]
IMO the fundamental problem here is we chose the wrong API for `export()`.

Today's `export()` is really doing logically two very different things:
1) Exporting the program
2) Normalizing the IR in one specific way, with lots of choices and assumptions baked in (optimized for inference use cases, and incompatible with inductor)

`export()` is too complicated and is trying to do two pretty different things at once.  Rather than trying to cram more stuff into `export()` (it is already too complicated), we should split it into two functions.  We should have a clean and simple `export()` that doesn't try to do any IR normalization.  Then have separate `normalize_for_X` or `optimize_for_Y` functions used when exporting for edge devices.

**Comment by gmagogsfm** [comment]
> IMO the fundamental problem here is we chose the wrong API for `export()`.
> 
> Today's `export()` is really doing logically two very different things:
> 
> 1. Exporting the program
> 2. Normalizing the IR in one specific way, with lots of choices and assumptions baked in (optimized for inference use cases, and incompatible with inductor)
> 
> `export()` is too complicated and is trying to do two pretty different things at once. Rather than trying to cram more stuff into `export()` (it is already too complicated), we should split it into two functions. We should have a clean and simple `export()` that doesn't try to do any IR normalization. Then have separate `normalize_for_X` or `optimize_for_Y` functions used when exporting for edge devices.

Agreed we have observed the same problem and there is ongoing work (some of them may land before PTC).

Today torch.export() does 1) capture program 2) runs aot dispatch 3) lowers to core ATen op set

We aim for following state:
torch.export.export() -> captures the program at pre-dispatch stage, without normalizaiong. 
torch.export.run_lowering() (name TBD) -> users chooses the configuration and decomp table they need, this is analogous to normalize_for_X().

But even after the change, it isn't planned to have a Torch IR stage, because of all the same reasons listed here:
https://github.com/pytorch/pytorch/blob/2c1554a0323107d821be3ff13df7833b9f0b960d/torch/_inductor/fx_passes/pre_grad.py#L55

**Comment by jansel** [comment]
I think we will hit the same exact problems with that plan and just repeat our past mistakes.  There will be lots of choices and assumptions baked into the pre-dispatch tracing, and different ways to configure it.

There is also the problem that it makes `export()` incompatible with `torch.compile()`/`torch.fx.symbolic_trace()`.  `export()` will produce a different IR than `compile()`, so all the 10-20 `torch.compile()` backends won't work out of the box with `export()`.    Similarly, all of the many FX passes in the ecosystem and internal (which are expecting to see modules in the graph) won't work.

Yes, Torch IR has issues.  Yes, we should migrate Inductor passes to pre-dispatch IR.  But, no we should not make `export()` incompatible with the rest of the ecosystem because of that.  We are required to support Torch IR basically forever because of `torch.fx.symbolic_trace`.

We should have `export()` produce Torch IR.  Then have a big comment in the docs saying "you should consider calling `normalize_for_x`, `normalize_for_y`, or `normalize_for_z`" (where one of the normalizations is pre_grad()) is strictly better.  All these normalizations throw away information.  You can always apply the normalizations later, but you can never un-apply them.  So the separate API is strictly more powerful than the combined one.

**Comment by gmagogsfm** [comment]
Closing this. @jansel and I discussed offline to eventually produce Torch IR from torch.export. (We may land at an intermediate solution for PTC to avoid BC breaking changes). We don't need to move pre grad passes to dynamo any more.
